server/api/auth/facebook.js

```diff
@@ -0,0 +1,61 @@
+const passport = require('passport');
+const passportFacebook = require('passport-facebook');
+const loginWithIdp = require('./loginWithIdp');
+
+const radix = 10;
+const PORT = parseInt(process.env.REACT_APP_DEV_API_SERVER_PORT, radix);
+const rootUrl = process.env.REACT_APP_CANONICAL_ROOT_URL;
+const clientID = process.env.REACT_APP_FACEBOOK_APP_ID;
+const clientSecret = process.env.FACEBOOK_APP_SECRET;
+
+const FacebookStrategy = passportFacebook.Strategy;
+let callbackURL = null;
+
+const useDevApiServer = process.env.NODE_ENV === 'development' && !!PORT;
+
+if (useDevApiServer) {
+  callbackURL = `http://localhost:${PORT}/api/auth/facebook/callback`;
+} else {
+  callbackURL = `${rootUrl}/api/auth/facebook/callback`;
+}
+
+const strategyOptions = {
+  clientID,
+  clientSecret,
+  callbackURL,
+  profileFields: ['id', 'name', 'emails'],
+  passReqToCallback: true,
+};
+
+const verifyCallback = (req, accessToken, refreshToken, profile, done) => {
+  const { email, first_name, last_name } = profile._json;
+
+  const userData = {
+    email,
+    firstName: first_name,
+    lastName: last_name,
+    accessToken,
+    refreshToken,
+    returnUrl: req.query.state ? req.query.state : null,
+  };
+
+  done(null, userData);
+};
+
+// ClientId is required when adding a new Facebook strategy to passport
+if (clientID) {
+  passport.use(new FacebookStrategy(strategyOptions, verifyCallback));
+}
+
+exports.authenticateFacebook = (req, res, next) => {
+  const from = req && req.query ? req.query.from : null;
+  passport.authenticate('facebook', { scope: ['email'], state: from })(req, res, next);
```

Comment:
How fixed this syntax is? 
I really think we should include other possible return paths as URL params too - and this "state" term is confusing. 

So, could this be something like:
```js
// pseudocode
const query = req && req.query;
const returnURL =  query && req.query.from 
  ? query.from 
  : query.defaultURL
  ? query.defaultURL 
  : null;
const confirmURL =  query && query.confirmURL ? query.confirmURL : null; 
passport.authenticate('facebook', { scope: ['email'], returnURL, confirmURL })(req, res, next);
```

Side note: I'm a bit confused about how these `from`/`state`/`user.returnUrl` work at the moment. Maybe some parameter renaming would help?